File: problems/0269-alien-dictionary/typescript/solution.ts
// Alien Dictionary — https://leetcode.com/problems/alien-dictionary/
export function alienOrder(words: string[]): string {
  const adj = new Map<string, Set<string>>();
  const indegree = new Map<string, number>();
  for (const w of words) {
    for (const c of w) {
      if (!adj.has(c)) {
        adj.set(c, new Set());
        indegree.set(c, 0);
      }
    }
  }

  for (let i = 0; i < words.length - 1; i++) {
    const w1 = words[i];
    const w2 = words[i + 1];
    const minLen = Math.min(w1.length, w2.length);
    if (w1.length > w2.length && w1.slice(0, minLen) === w2.slice(0, minLen)) {
      return ""; // invalid prefix ordering
    }
    for (let j = 0; j < minLen; j++) {
      if (w1[j] !== w2[j]) {
        const set = adj.get(w1[j]);
        if (set !== undefined && !set.has(w2[j])) {
          set.add(w2[j]);
          indegree.set(w2[j], (indegree.get(w2[j]) ?? 0) + 1);
        }
        break;
      }
    }
  }

  const ready: string[] = [];
  for (const [c, d] of indegree) if (d === 0) ready.push(c);
  const res: string[] = [];
  while (ready.length > 0) {
    ready.sort(); // smallest-letter-first for determinism
    const c = ready.shift();
    if (c === undefined) break;
    res.push(c);
    for (const nb of adj.get(c) ?? []) {
      const d = (indegree.get(nb) ?? 0) - 1;
      indegree.set(nb, d);
      if (d === 0) ready.push(nb);
    }
  }

  return res.length === indegree.size ? res.join("") : "";
}

File: problems/0269-alien-dictionary/go/solution.go
package solution

import "sort"

// Alien Dictionary — https://leetcode.com/problems/alien-dictionary/
func alienOrder(words []string) string {
	adj := map[byte]map[byte]bool{}
	indegree := map[byte]int{}
	for _, w := range words {
		for i := 0; i < len(w); i++ {
			if _, ok := adj[w[i]]; !ok {
				adj[w[i]] = map[byte]bool{}
				indegree[w[i]] = 0
			}
		}
	}

	for i := 0; i < len(words)-1; i++ {
		w1, w2 := words[i], words[i+1]
		minLen := len(w1)
		if len(w2) < minLen {
			minLen = len(w2)
		}
		if len(w1) > len(w2) && w1[:minLen] == w2[:minLen] {
			return "" // invalid prefix ordering
		}
		for j := 0; j < minLen; j++ {
			if w1[j] != w2[j] {
				if !adj[w1[j]][w2[j]] {
					adj[w1[j]][w2[j]] = true
					indegree[w2[j]]++
				}
				break
			}
		}
	}

	ready := []byte{}
	for c := range indegree {
		if indegree[c] == 0 {
			ready = append(ready, c)
		}
	}
	res := []byte{}
	for len(ready) > 0 {
		sort.Slice(ready, func(i, j int) bool { return ready[i] < ready[j] }) // smallest-first
		c := ready[0]
		ready = ready[1:]
		res = append(res, c)
		for nb := range adj[c] {
			indegree[nb]--
			if indegree[nb] == 0 {
				ready = append(ready, nb)
			}
		}
	}

	if len(res) == len(indegree) {
		return string(res)
	}
	return ""
}
